extension/src/CSV/ImportXliffCSV.ts

```diff
@@ -1,32 +1,75 @@
 
-export function importXliffCSV(updateXlf: Xliff, csvPath: string): number {
+export function importXliffCSV(updateXlf: Xliff, csvPath: string, useExternalTranslationTool: boolean, xliffCSVImportTargetState: string): number {
+    let requiredHeaders: string[] = ["Id", "Source", "Target"];
+
     let updatedTargets: number = 0;
     let csv = new CSV();
     csv.encoding = "utf8bom";
     csv.readFileSync(csvPath);
-    testRequiredHeaders(csv.headers);
+    let updateTargetState = false;
+    let updateTargetStateFromCsv = false;
+    let newTargetState: TargetState;
+    if (useExternalTranslationTool) {
+        switch (xliffCSVImportTargetState.toLowerCase()) {
+            case "(leave)":
+                updateTargetState = false;
+                break;
+            case "(from csv)":
+                updateTargetState = true;
+                updateTargetStateFromCsv = true;
+                break;
+            default:
+                updateTargetState = true;
+                newTargetState = <TargetState>xliffCSVImportTargetState;
+                break;
+        }
+    }
+    if (updateTargetStateFromCsv) {
+        requiredHeaders.push("State");
+    }
+    const headerIndexMap = csv.headerIndexMap;
+    testRequiredHeaders(headerIndexMap, requiredHeaders);
+
     csv.lines.filter(l => l.length > 1).forEach(line => {
-        let values = { id: line[0], source: line[1], target: line[2] }
-        let transunit = updateXlf.getTransUnitById(values.id);
-        if (transunit.source !== values.source) {
-            throw new Error(`Sources doesn't match for id ${transunit.id}.\nExisting Source: "${transunit.source}".\nImported source: "${values.source}"`);
+        let values = { id: line[<number>headerIndexMap.get("Id")], source: line[<number>headerIndexMap.get("Source")], target: line[<number>headerIndexMap.get("Target")], state: "" }
+        if (updateTargetStateFromCsv) {
+            values.state = line[<number>headerIndexMap.get("State")];
         }
-        if (transunit.targetTextContent !== values.target) {
-            transunit.targets[0].textContent = values.target;
+
+        let transUnit = updateXlf.getTransUnitById(values.id);
+        if (isNullOrUndefined(transUnit)) {
+            throw new Error(`Could not find any translation unit with id "${values.id}" in "${updateXlf._path}"`);
+        }
+        if (transUnit.source !== values.source) {
+            throw new Error(`Sources doesn't match for id ${transUnit.id}.\nExisting Source: "${transUnit.source}".\nImported source: "${values.source}"`);
+        }
+
+        if (transUnit.target.textContent !== values.target) {
+            transUnit.target.textContent = values.target;
+            if (updateTargetState) {
+                if (updateTargetStateFromCsv) {
+                    transUnit.target.state = <TargetState>values.state;
+                } else {
+                    transUnit.target.state = newTargetState;
+                }
```

Comment:
Ternay operator could be nice.
```suggestion
               transUnit.target.state = updateTargetStateFromCsv ? values.state as TargetState : newTargetState;
```

Reply:
I will probably never get used to ternary operators... 😉

Reply:
I've found my self wanting to write it in AL :s

Reply:
I would never use if they introduced it... 😜